csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedElement.qll

```diff
@@ -0,0 +1,563 @@
+  //    // its arguments, though, because we use them as part of the synthesis.
+  //    newExpr.getAllocatorCall() = expr
+  //  ) or
+  not translateFunction(expr.getEnclosingCallable())
+  or
+  // Ignore size of arrays when translating
+  (expr.getParent() instanceof ArrayCreation and expr.hasValue()) or
+  // Ignore the child expression of a goto case stmt
+  (expr.getParent() instanceof GotoCaseStmt)
+}
+
+/**
+ * Holds if `expr` should be ignored for the purposes of IR generation.
+ */
+private predicate ignoreExpr(Expr expr) {
+  ignoreExprOnly(expr) or
+  ignoreExprAndDescendants(expr)
+}
+
+// TODO: Probably ok to ignore for now
+///**
+// * Holds if `func` contains an AST that cannot be translated into IR. This is mostly used to work
+// * around extractor bugs. Once the relevant extractor bugs are fixed, this predicate can be removed.
+// */
+//private predicate isInvalidFunction(Callable callable) {
+//  exists(Literal literal |
+//    // Constructor field inits within a compiler-generated copy constructor have a source expression
+//    // that is a `Literal` with no value.
+//    literal = callable.(Constructor).getInitializer().(ConstructorFieldInit).getExpr() and
+//    not exists(literal.getValue())
+//  ) or
+//  exists(ThisExpr thisExpr |
+//    // An instantiation of a member function template is not treated as a `MemberFunction` if it has
+//    // only non-type template arguments.
+//    thisExpr.getEnclosingFunction() = callable and
+//    not callable instanceof Member
+//  ) or
+//  exists(Expr expr |
+//    // Expression missing a type.
+//    expr.getEnclosingFunction() = callable and
+//    not exists(expr.getType())
+//  )
+//}
+/**
+ * Holds if `func` should be translated to IR.
+ */
+private predicate translateFunction(Callable callable) {
+  // not isInvalidFunction(callable)
+  exists(callable.getEntryPoint()) and
+  exists(IRConfiguration config | config.shouldCreateIRForFunction(callable))
+}
+
+/**
+ * Holds if `stmt` should be translated to IR.
+ */
+private predicate translateStmt(Stmt stmt) { translateFunction(stmt.getEnclosingCallable()) }
+
+/**
+ * Holds if `expr` is most naturally evaluated as control flow, rather than as
+ * a value.
+ */
+private predicate isNativeCondition(Expr expr) {
+  expr instanceof BinaryLogicalOperation and
+  not isIRConstant(expr)
+}
+
+/**
+ * Holds if `expr` can be evaluated as either a condition or a value expression,
+ * depending on context.
+ */
+private predicate isFlexibleCondition(Expr expr) {
+  (
+    expr instanceof ParenthesizedExpr or
+    expr instanceof LogicalNotExpr
+  ) and
+  usedAsCondition(expr) and
+  not isIRConstant(expr)
+}
+
+/**
+ * Holds if `expr` is used in a condition context, i.e. the Boolean result of
+ * the expression is directly used to determine control flow.
+ */
+private predicate usedAsCondition(Expr expr) {
+  exists(BinaryLogicalOperation op |
+    op.getLeftOperand() = expr or
+    op.getRightOperand() = expr
```

Comment:
Since the expression can either be the left or right operand and we only care if it is used in a condition. I suppose `getAnOperand` could have been used here.

Reply:
Ok, I misunderstood. Yes, `getAnOperand()` seems nicer.